clients/pkg/logentry/stages/limit.go

```diff
@@ -0,0 +1,202 @@
+package stages
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"regexp"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/time/rate"
+)
+
+const (
+	ErrLimitStageEmptyConfig        = "limit stage config must contain at least one of `source`, `expression`, `older_than` or `longer_than`"
+	ErrLimitStageInvalidConfig      = "limit stage config error, `value` and `expression` cannot both be defined at the same time."
+	ErrLimitStageInvalidRegex       = "limit stage regex compilation error: %v"
+	ErrLimitStageInvalidRateOrBurst = "limit stage failed to parse rate or burst"
+)
+
+var ratelimitDropReason = "ratelimit_drop_stage"
+
+type LimitConfig struct {
+	Source     *string `mapstructure:"source"`
+	Rate       float64 `mapstructure:"rate"`
+	Burst      int     `mapstructure:"burst"`
+	Drop       bool    `mapstructure:"drop"`
+	Value      *string `mapstructure:"value"`
+	Expression *string `mapstructure:"expression"`
```

Comment:
Why do we need Value, Expression and Source here ? We could easily combine this stage with a `match` stage no ? I wonder if we need this logic in the limit stage ? see https://grafana.com/docs/loki/latest/clients/promtail/stages/match/

```yaml
match:
  selector: '{app="foo"} |~ "regex"'
  stages:
  - limit:
      rate: 10
      drop: true
```

@grafana/loki-team WDYT ? The only difference is that you can't use an extracted value, but regex goes already a long way.

Reply:
Yeah, I think using the existing `match` stage for this makes sense to me.

Reply:
Ok let's remove all this logic, if we have a request later we can think about how to do that. I want to keep each stage responsibility small and simple.